blog/_posts/2019-01-18-fluxdiffeq.md

```diff
@@ -0,0 +1,742 @@
+kind of equation is known as a stochastic differential equation (SDE).
+Since [DifferentialEquations.jl handles SDEs](http://docs.juliadiffeq.org/latest/tutorials/sde_example.html)
+(and is currently the only library with adaptive stiff and non-stiff SDE integrators),
+these can be handled as a layer in Flux similarly. Here's a neural net layer
+with an SDE:
+
+```julia
+function lotka_volterra_noise(du,u,p,t)
+  du[1] = 0.1u[1]
+  du[2] = 0.1u[2]
+end
+prob = SDEProblem(lotka_volterra,lotka_volterra_noise,[1.0,1.0],(0.0,10.0))
+
+p = param([2.2, 1.0, 2.0, 0.4])
+params = Flux.Params([p])
+function predict_fd_sde()
+  diffeq_fd(p,reduction,101,prob,SOSRI(),saveat=0.1)
+end
+loss_fd_sde() = sum(abs2,x-1 for x in predict_fd_sde())
+loss_fd_sde()
+```
+
+And we can train the neural net to watch it in action and find parameters to make
+the amount of bunnies close to constant:
+
+```julia
+data = Iterators.repeated((), 100)
+opt = ADAM(0.1)
+cb = function ()
+  display(loss_fd_sde())
+  display(plot(solve(remake(prob,p=Flux.data(p)),SOSRI(),saveat=0.1),ylim=(0,6)))
+end
+
+# Display the ODE with the current parameter values.
+cb()
+
+Flux.train!(loss_fd_sde, params, data, opt, cb = cb)
+```
+
+![SDE NN Animation](https://user-images.githubusercontent.com/1814174/51399524-2c6abf80-1b14-11e9-96ae-0192f7debd03.gif)
+
+And we can keep going. There are differential equations
+[which are piecewise constant](http://docs.juliadiffeq.org/latest/tutorials/discrete_stochastic_example.html)
+used in biological simulations, or
+[jump diffusion equations from financial models](http://docs.juliadiffeq.org/latest/tutorials/jump_diffusion.html),
+and the solvers map right over to the Flux neural network frame work through FluxDiffEq.jl
+FluxDiffEq.jl uses only around ~100 lines of code to pull this all off.
+
+## Implementing the Neural ODE layer in Julia
+
+Let's go all the way back for a second and now implement the neural ODE layer
+in Julia. Remember that this is simply an ODE where the derivative function
+is defined by a neural network itself. To do this, let's first define the
+neural net for the derivative. In Flux, we can define a multilayer perceptron
+with 1 hidden layer and a `tanh` activation function like:
+
+```julia
+dudt = Chain(Dense(2,50,tanh),Dense(50,2))
+```
+
+To define a `neural_ode` layer, we then just need to give
+it a timespan and use the `neural_ode` function:
+
+```julia
+tspan = (0.0f0,25.0f0)
+x->neural_ode(dudt,x,tspan,Tsit5(),saveat=0.1)
+```
+
+As a side note, to run this on the GPU, it is sufficient to make the initial
+condition and neural network be on the GPU. This will cause the entire ODE
+solver's internal operations to take place on the GPU without extra data
+transfers in the integration scheme. This looks like[^gpu]:
+
+```julia
+x->neural_ode(gpu(dudt),gpu(x),tspan,BS3(),saveat=0.1)
+```
+
+[^gpu]: We have changed the solver from Tsit5 to BS3, as Tsit5 is currently not GPU compatible.

```

Comment:
That's why I said "currently". Maybe we should add the link over [currently not GPU compatible] so it's more heavily implied.

Reply:
This is a delicate subject because then it sounds like it's hard for us to make this GPU compatible. It isn't, we already did it, and we had to revert the PR since v1.0 broadcasting is 3x slower (https://github.com/JuliaLang/julia/issues/28126) and users immediately reported the regression. So we are purposefully choosing to not support GPUs on this algorithm right now because of that, BUT we could add it any time in less than 20 minutes since the code with broadcast is already there and commented out.

How do we politely say this in a way that conveys "okay yes, they essentially have GPU compatibility on everything".

Reply:
``
```suggestion
[^gpu]: We have changed the solver from Tsit5 to BS3, as Tsit5 is [currently not GPU compatible]((https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/106)).
```